### DISCUSSION TITLE
How to build SQL query with jsonb_build_object() and jsonb_agg() Postgres functions on the Supabase JS client?

### DISCUSSION THREAD

**Discussion by koddr** [supabase/supabase#3761] [Questions] [answered]
Hi,

I'm transferring my project from a usual database to Supabase immediately. And the problem is that this SQL query is simply not clear how I can write on the Supabase JavaScript client (`@supabase/supabase-js`):

> Unfortunately, I did not find an answer to this question in the [Fetch data: select()](https://supabase.io/docs/reference/javascript/select) documentation. 

```sql
SELECT
	p.id,
	jsonb_build_object('user_id', u.id) AS author,
	COUNT(t.id) AS tasks_count,
	COALESCE(
		jsonb_agg(
			jsonb_build_object('id', t.id, 'status', t.status)  
		)
		FILTER (WHERE t.project_id IS NOT NULL), '[]'
	) AS tasks
FROM
	projects AS p
	LEFT JOIN users AS u ON u.id = p.user_id
	LEFT JOIN tasks AS t ON t.project_id = p.id
GROUP BY
	p.id,
	u.id
LIMIT 1
```

That is, I need to output a project (from the `projects` table) that has two JSONB fields (_generated on the fly_), which are assembled from linked tables `users` (default Supabase table for users) and `tasks`.

- `projects` table has FK to `users` table (`projects.user_id` -> `users.id`)
- `tasks` table has FK to `projects` table (`tasks.project_id` -> `projects.id`)

This SQL query in your query interface works fine:

![Screenshot 2021-11-03 at 09 13 24](https://user-images.githubusercontent.com/11155743/140016446-a6adffd2-fb61-4dbf-a4cd-1d878153563b.png)

But how do I transfer such a complex query to the realities of your JS client?

> Perhaps it would be cool to have something like a “**raw SQL query**” to just insert the whole query like this “as is” (which works fine in the Supabase control panel). Something like this:
>
>```js
>const { data, error } = await supabase.sql.raw(`
>  SELECT
>    p.id,
>    jsonb_build_object('user_id', u.id) AS author,
>    COUNT(t.id) AS tasks_count,
>    ...
>    ...
>`)
>```

Very much counting on your help and support, as this is currently the last stop factor for moving to Supabase for my project.

Thank you in advance for your help! 😉

**Comment by dragarcia** [comment, marked as answer]
Hi @koddr 👋 

If this is a query that you would be using often, I think the usage of views might be a possible option for you. Following this [guide](https://supabase.io/blog/2020/11/18/postgresql-views), you can set up the view (let's call it `projects_enhanced`) through the dashboard accordingly:
```sql
CREATE VIEW projects_enhanced AS
    SELECT
	     p.id,
	    jsonb_build_object('user_id', u.id) AS author,
	    COUNT(t.id) AS tasks_count,
	    COALESCE(
		    jsonb_agg(
			    jsonb_build_object('id', t.id, 'status', t.status)  
		    )
		    FILTER (WHERE t.project_id IS NOT NULL), '[]'
	    ) AS tasks
    FROM
	    projects AS p
	    LEFT JOIN users AS u ON u.id = p.user_id
	    LEFT JOIN tasks AS t ON t.project_id = p.id
    GROUP BY
	    p.id,
	    u.id
    LIMIT 1
```

Afterward, from the supabase-js library, you can then call this view as such:
```js
const { data: projects, error } = await supabase
    .from('projects_enhanced')
    .select('*')
```

**Reply by koddr** [reply]
Thanks, @dragarcia!

Wow, that turned out to be easier than I thought! 👍 

> I'll have to do a PR with an example for the documentation, since it's not at all obvious when working with the Supabase interface.

As a side note, where can I see the list of `Postgres View` created this way? It would be great to have a separate menu item in Supabase admin panel for the View created, so I can go in and change it at any time.

**Reply by dragarcia** [reply]
You should be able to see the content of any created views on the left hand side of the Table Editor of the dashboard. Under `All Tables` there should be a section labelled `All Views` right underneath if you have successfully created one. Unfortunately, if you want to modify the underlying query of the view, you would still have to do it over the SQL Editor of the dashboard using [ALTER VIEW](https://www.postgresql.org/docs/13/sql-alterview.html). To see the current underlying query of the view, you can execute the following:
```sql
SELECT pg_get_viewdef('view_name', true)
```

**Reply by koddr** [reply]
@dragarcia thank you! 😉
